components/raftstore/src/store/peer.rs

```diff
@@ -704,6 +726,326 @@ pub enum UnsafeRecoveryState {
+            "peer_id" => self.peer_id(),
+            "msg_type" => ?ty,
+            "to" => to.get_id()
+        );
+        send_msg.set_extra_msg(msg);
+        send_msg.set_to_peer(to.clone());
+        if let Err(e) = trans.send(send_msg) {
+            error!(?e;
+                "failed to send extra message";
+                "type" => ?ty,
+                "region_id" => self.region_id(),
+                "peer_id" => self.peer_id(),
+                "target" => ?to,
+            );
+        }
+    }
+
+    fn prepare_raft_message(&self) -> RaftMessage {
+        let mut send_msg = RaftMessage::default();
+        send_msg.set_region_id(self.region_id());
+        // set current epoch
+        send_msg.set_region_epoch(self.region().get_region_epoch().clone());
+        send_msg.set_from_peer(self.peer().clone());
+        send_msg
+    }
+
+    // When a replica cannot detect any leader, `MsgReadIndex` will be dropped,
+    // which would cause a long time waiting for a read response. Then we should
+    // return an error directly in this situation.
+    // return true if callback should be called
+    fn read_index_no_leader<T: Transport>(
+        &mut self,
+        trans: &mut T,
+        pd_scheduler: &mut Scheduler<PdTask<EK, ER>>,
+        err_resp: &mut RaftCmdResponse,
+    ) -> bool {
+        if !self.is_leader() && self.leader_id() == INVALID_ID {
+            // The leader may be hibernated, send a message for trying to awaken the leader.
+            if self.mut_bcast_wake_up_time().is_none()
+                || self
+                    .mut_bcast_wake_up_time()
+                    .as_ref()
+                    .unwrap()
+                    .saturating_elapsed()
+                    >= Duration::from_millis(MIN_BCAST_WAKE_UP_INTERVAL)
+            {
+                self.bcast_wake_up_message(trans);
+                self.mut_bcast_wake_up_time()
+                    .replace(TiInstant::now_coarse());
+
+                let task = PdTask::QueryRegionLeader {
+                    region_id: self.region_id(),
+                };
+                if let Err(e) = pd_scheduler.schedule(task) {
+                    error!(
+                        "failed to notify pd";
+                        "region_id" => self.region_id(),
+                        "peer_id" => self.peer_id(),
+                        "err" => %e,
+                    )
+                }
+            }
+            self.set_should_wake_up(true);
+            cmd_resp::bind_error(err_resp, Error::NotLeader(self.region_id(), None));
+            return true;
+        }
+        false
+    }
+
+    /// Try to renew leader lease.
+    fn maybe_renew_leader_lease(
+        &mut self,
+        ts: Timespec,
+        store_meta: &mut Arc<Mutex<StoreMeta>>,
+        progress: Option<ReadProgress>,
+    ) {
+        // A nonleader peer should never has leader lease.
+        let read_progress = if !self.is_leader() {
```

Comment:
I suggest to integrate them with `Lease` struct, so all the checks can be hidden, and probably be more efficient because no need to check every condition but just a bool flag like:
```
fn on_propose_split() {
    self.lease.prevent_renew();
}

fn maybe_renew_leader_lease() {
    if let Some(lease) = self.lease.renew() {
       ...
   }
}
```

Reply:
I don't think it has big advantage, if it has any advantage at all. Because you have to introduce these hook APIs like on_propose_split, on_propose_merge, and on_force_leader and call them properly. if it missed any call, then it will be a bug. And what's the return of this change?  It reduces two bool comparison! Is it a big deal?